### ISSUE TITLE
Upload from a URL to cloud

### ISSUE THREAD

**Issue by ntbatch** [ncw/rclone#3282]
#### What is your current rclone version (output from `rclone version`)?
rclone v1.48.0
- os/arch: android/arm
- go version: go1.12.6

#### What problem are you are trying to solve?
`rclone copyto http://url/path/file.ext remote:path/file.ext`
or new command "copyurl"
`rclone copyurl http://url/path/file.ext remote:path/[file.ext]`

#### How do you think rclone should be changed to solve that?
Dropbox, Yandex and OneDrive(personal) have this function.

https://www.dropbox.com/developers/documentation/http/documentation#files-save_url

https://tech.yandex.com/disk/api/reference/upload-ext-docpage/

https://docs.microsoft.com/en-us/onedrive/developer/rest-api/api/driveitem_upload_url?view=odsp-graph-online#http-request

Maybe some other clouds...

**Comment by ncw** [comment]
So do a server side copy from a URL?

This would be possible to implement - `rclone copyurl` could use the server side option if available

Do you want to work on this?

**Comment by ntbatch** [comment]
> So do a server side copy from a URL?

Yes, it's server side copy. It really work. Anyone can try run examples from dropbox page with curl.

`curl -q -X POST https://api.dropboxapi.com/2/files/save_url --header "Authorization: Bearer YOUR_DROPBOX_TOKEN_HERE" --header "Content-Type: application/json" --data "{\"path\": \"/blabla.exe\",\"url\": \"https://rarlab.com/rar/wrar571.exe\"}"`

`curl -q -X POST https://api.dropboxapi.com/2/files/save_url --header "Authorization: Bearer YOUR_DROPBOX_TOKEN_HERE" --header "Content-Type: application/json" --data "{\"path\": \"/blabla.ogg\",\"url\": \"https://ftp.gnu.org/video/lp2010-sat-eben-moglen.ogg\"}"`

`curl -q -X POST https://api.dropboxapi.com/2/files/save_url --header "Authorization: Bearer YOUR_DROPBOX_TOKEN_HERE" --header "Content-Type: application/json" --data "{\"path\": \"/blabla.iso\",\"url\": \"https://cdimage.debian.org/debian-cd/current/i386/iso-cd/debian-9.9.0-i386-netinst.iso\"}"`

Curl output will be like "{".tag": "async_job_id", "async_job_id": "fbv5Q2OrFU0AAAAAAAAPEg"}"

And can check after 1-5 min upload status:
`curl -q -X POST https://api.dropboxapi.com/2/files/save_url/check_job_status --header "Authorization: Bearer YOUR_DROPBOX_TOKEN_HERE" --header "Content-Type: application/json" --data "{\"async_job_id\": \"fbv5Q2OrFU0AAAAAAAAPEg\"}"`

Curl output will be "in_progress", "complete" or something another, e.g. "failed". I tried upload via curl command to API files about 100mb - 1.37gb. Okay. File 2.9gb  failed(timeout).

API probably has some limitations, time(in DB about 5 min), job count, i don't know.

>Do you want to work on this?

I'm not programner, just user. Can be beta-tester only :-)
I just found this theme:
https://forum.rclone.org/t/direct-cloud-to-cloud-copy/7219/5 and read your last post.
Sometimes command "copyurl" would be very useful. Especially on the smartphone.

**Comment by ncw** [comment]
To implement this we need a new optional interface for rclone, say `CopyURL(remote, URL string) error`.

Backends that could implement this would do the necessary to get the remote cloud storage to fetch the URL.

The command `rclone copyurl` would
- if `CopyURL` was implemented then call it and do the fetch remotely
- download and reupload the data if not.
- maybe add an `--async` flag so rclone doesn't wait for completion.

> I'm not programmer, just user. Can be beta-tester only :-)

If you wanted to help, finding out which of rclone's backends could support this and linking to the documentation where it is described would be really useful :-)

**Comment by varenc** [comment]
I'm interested in this!

@ncw do you think this would be a reasonable feature for a new rclone contributor to add?

Here's the docs on some of the backends that support this:
- put.io: https://app.swaggerhub.com/apis-docs/putio/putio/2.7.0#/transfers/post_transfers_add  
- Dropbox: https://www.dropbox.com/developers/documentation/http/documentation#files-save_url
- OneDrive: https://docs.microsoft.com/en-us/onedrive/developer/rest-api/api/driveitem_upload_url?view=odsp-graph-online  (personal accounts only)
- Yandex: https://tech.yandex.com/disk/api/reference/upload-ext-docpage/
- Google Cloud Storage: Doesn't really support it out of the box, but the additional "Cloud Storage Transfer Service" supports this in a more batch-oriented way: https://cloud.google.com/storage-transfer/docs/reference/rest/v1/TransferSpec#httpdata

Backends that I checked don't support this: Box, S3, Google Drive,  Backblaze, [Azure Blob Storage](https://docs.microsoft.com/en-us/rest/api/storageservices/copy-blob-from-url) (seems to only allow URLs that point to other things in blob storage)


For the rclone implementation:
- Would it be reasonable to just implement this for one backend first?  
- With the new `--async` option, would rclone just initiate the transfer, confirm it started successfully, and then quit?  And not stay running and constantly check the progress?  That would certainly be simpler!  The downside of that approach is that if the transfer fails after it's been started there's no feedback for the user. I could make it so each implementation does a few quick sanity checks to make sure the transfer started and didn't immediately error out though.  (though on the dropbox backend, for example, I know save_url requests get put in a queue and it might take a minute for the transfer to get popped from the queue and start.  If the URL is invalid for some reason you won't be able to see that it errored out until after the job pops from the queue)
- Also, the flag `--async` to me doesn't sound like it necessarily means this is a Cloud=>Clould transfer with no local bandwidth usage. Maybe `--async-cloud` or `--async-direct` could be better (though don't love those names)



(*digression*:  Some backends also allow you to create download URLs for individual files.  Or sometimes they allow this implicitly in weird ways.  With this functionality, it coul be possible so that when you use `rclone copy --async` between two different supported cloud backends it could do a direct "Cloud to Cloud" transfer where none of the files are downloaded/uploaded locally!  But that has a bunch more challenges I'm sure... first step is making this work with `copyurl`)

**Comment by ncw** [comment]
> I'm interested in this!

Great!

> do you think this would be a reasonable feature for a new rclone contributor to add?

Absolutely :-)

> Here's the docs on some of the backends that support this:
[snip excellent research]

> For the rclone implementation:
> 
> * Would it be reasonable to just implement this for one backend first?

I would start with one to get the interface working. When everything was working then I'd try to implement as many backends as possible.  I try to avoid features with only one implementation - the "orphaned" feature problem!

> * With the new `--async` option, would rclone just initiate the transfer, confirm it started successfully, and then quit?

Yes, that was the idea.

>  And not stay running and constantly check the progress?

That is what rclone would do without the `--async` flag

>  That would certainly be simpler!  The downside of that approach is that if the transfer fails after it's been started there's no feedback for the user.

That is what you want when you don't provide the `--async` flag.

If I had to choose one of other behaviour I'd choose without the `--async` flag .

> * Also, the flag `--async` to me doesn't sound like it necessarily means this is a Cloud=>Clould transfer with no local bandwidth usage. Maybe `--async-cloud` or `--async-direct` could be better (though don't love those names)

Async meaning rclone isn't going to wait for the result but return immediately.
 
> (_digression_: Some backends also allow you to create download URLs for individual files. Or sometimes they allow this implicitly in weird ways. With this functionality, it coul be possible so that when you use `rclone copy --async` between two different supported cloud backends it could do a direct "Cloud to Cloud" transfer where none of the files are downloaded/uploaded locally! But that has a bunch more challenges I'm sure... first step is making this work with `copyurl`)

Yes that is a good idea and would be a great feature for rclone.... for later :-)

I think the first thing to think about would be the interface for `CopyURL`

This interface would work for the `synchronous` copy which should be the default.

```go
type CopyURLer interface {
    // CopyURL fetches the data at URL and stores it in an object at remote
    // The data is copied direct by the cloud storage backend, not via rclone
    // if err != nil then obj will point to the created object
    CopyURL(remote, URL string) (obj Object, err error)
}
```

Thinking some more I think the async interface might be a mistake.  Rclone users appreciate data integrity and it is going to be difficult to ensure that with async. If we want to implement it then the above interface could have an `async` flag.  However you can be sure the next user request will be to list in progress transfers, see their progress, cancel them etc!

Speaking of progress, the backend copy could attempt to report progress - I don't know whether the backends support that, so maybe it is best not to complicate the interface.

**Comment by jrochkind** [comment]
Azure seems to now supports a method of server-side copy from URL  and the `azcopy` tool uses it for copies from AWS S3:

> AzCopy copies data from AWS S3 with high throughput by scaling out copy jobs to multiple Azure Storage servers. AzCopy relies on the new Azure Storage REST API operation Put Block from URL, which copies data directly from a given URL. Using Put Block from URL, AzCopy v10 moves data from an AWS S3 bucket to an Azure Storage account, without first copying the data to the client machine where AzCopy is running. Instead, Azure Storage performs the copy operation directly from the source. Thanks to this method, the client in the middle is no longer the bottleneck.

—https://azure.microsoft.com/en-us/blog/move-your-data-from-aws-s3-to-azure-storage-using-azcopy/

The actual Azure API's involve seem kind of convoluted, they aren't just "give it a url" I don't think. I'm not certain of the details. 

But while the `azcopy` tool only supports copy from S3 this way... it seems like it should be possible to copy from any URL this way. 

It would be amazing if the rclone tool could allow server-side copy from a URL for more sources to azure than azcopy supports. (AZcopy currently doesn't even support other S3-compatible providers. https://github.com/Azure/azure-storage-azcopy/issues/832)

**Comment by ncw** [comment]
[Put block from URL](https://docs.microsoft.com/en-us/rest/api/storageservices/put-block-from-url) as the underlying API looks perfect for azure.

You could do a server side copy from S3 very easily with this by creating temporary download URLs for the S3 data and feeding it to the `CopyURL` method in the azure backend.

That kind of implies another way of doing a cross provider transfer. If the source supports `PublicLink` (`rclone link` command) and the destination support `CopyURL` (what this issue is about) then rclone could do a cross provider transfer.

There are lots of annoying details about modification times and hashes which might make this less than perfect, but in some situations not having the data pass through the computer that runs rclone is a big win.

**Comment by aviv926** [comment]
Is this something expected to be added to the rclone copyurl feature?

**Comment by ncw** [comment]
> Is this something expected to be added to the rclone copyurl feature?

Yes. The idea would be that if a backend implemented the `CopyURL` feature then it could copy straight from the URL that rclone provided.

@aviv926 are you interested in implementing this?

**Comment by maxgreen01** [comment]
Is this issue still open? I'd be interested in giving this a shot if so!

**Comment by ncw** [comment]
@maxgreen01 yes, still open. Not sure which backends can support it but quite a few do. Go ahead!

**Comment by maxgreen01** [comment]
Just wanted to give a brief update to anyone else who sees this that I've decided to work on a different rclone issue, so this one is still up for grabs!
